File: solutions/javascript/bank-account/1/bank-account.js
export class BankAccount {
  #value;
  #isOpen;
  
  constructor() {
    this.#isOpen = false;
  }

  open() {
    if (this.#isOpen) throw new ValueError;
    this.#isOpen = true;
    this.#value = 0;
  }

  close() {
    if (!this.#isOpen) throw new ValueError;
    this.#isOpen = false;
    this.#value = 0;
  }

  deposit(value) {
    if (!this.#isOpen || value < 0) throw new ValueError;
    this.#value += value;
  }

  withdraw(value) {
    if (!this.#isOpen || value < 0 || value > this.#value) throw new ValueError;
    this.#value -= value;
  }

  get balance() {
    if (!this.#isOpen) throw new ValueError;
    return this.#value;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}

File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"

#include <stdexcept>

namespace Bankaccount {

    auto Bankaccount::open() -> void {

        const std::lock_guard<std::mutex> lock(account_mutex);
        
        if (is_open)
        {
            throw std::runtime_error("can't open an account already opened.");
        }

        is_open = true;
        m_balance = 0;
    }

    auto Bankaccount::close() -> void {

        const std::lock_guard<std::mutex> lock(account_mutex);

        if (not is_open)
        {
            throw std::runtime_error("can't close an account not opened.");
        }

        is_open = false;
        m_balance = 0;
    }

    auto Bankaccount::deposit(double value) -> void {

        const std::lock_guard<std::mutex> lock(account_mutex);
             
        if (not is_open)
        {
            throw std::runtime_error("can't deposit to a not opened account.");
        }

        if (value < 0)
        {
            throw std::runtime_error("can't deposit a negative value.");
        }

        m_balance += value;
    }

    auto Bankaccount::withdraw(double value) -> void {

        const std::lock_guard<std::mutex> lock(account_mutex);

        if (not is_open)
        {
            throw std::runtime_error("can't withdraw from an account not opened.");
        }

        if (value > m_balance)
        {
            throw std::runtime_error("can't withdraw more than balance in account.");
        }

        if (value < 0)
        {
            throw std::runtime_error("can't withdraw a negative amount.");
        }

        m_balance -= value;
    }


    auto Bankaccount::balance() const -> double {

        if (not is_open)
        {
            throw std::runtime_error("can't check balance from an account not opened.");
        }
        
        return m_balance;
    }
}